### PR TITLE
ModBus RTU

### PR DESCRIPTION
**Pull request by theltybc**
### Contribution description
Implementation ModBus RTU

### Testing procedure
Example `examples/modbus`. 
Print in STDOUT note about requet to slave. And check result.

### Issues/PRs references
Feature Request: #13869

### CODE REVIEW

**Comment by theltybc** [comment]
Maybe need only Read(0x3) and Write(0x10) Multiple registers function? 
And possibele remove other functions as useless?

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ */
+typedef struct {
+  /**
+   * @brief  UART port.
+   */
+  uart_t uart;
+  /**
+   * @brief  baudrate.
+   */
+  uint32_t baudrate;
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:71]: Is this needed? Wouldn't it be sufficient to just initialize the UART interface with the correct symbol rate?

Also: "baud" is the unit for the symbol rate, (symbols per second). Thus "baudrate" is technically not correct, even though it is widely used. So if this is actually needed, I suggested to call it either `baud` or `symbolrate`.

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  /**
+   * @brief  RTS pin. 0 - disable.
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:73]: What does "0 - disable" mean?

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+   */
+  gpio_t pin_rts;
+  /**
+   * @brief  RTS signals when transmit.
+   */
+  int pin_rts_enable;
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:79]: Would setting `pin_rts` to `GPIO_UNDEF` also work for disabling RTS. If so, this would save 2-4 bytes, depending on architecture.

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  /**
+   * @brief  ID device. 0 - master, otherwice slave.
+   */
+  uint8_t id;
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:83]: You could safe a couple of bytes by reordering the members in the struct so that fewer padding bytes are needed. E.g. if you order members by alignment requirement, you would end up with as few padding bytes added as absolutely needed.

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  /**
+   * @brief   Before slave begin to response; usec.
+   */
+  uint32_t timeout;
+  /**
+   * @brief   Template buffer for receive message or send. **Must never be changed
+   *          by the user.**
+   * @internal
+   */
+  uint8_t _buffer[MODBUS_RTU_PACKET_SIZE_MAX];
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:93]: I don't think that you need to add an underscore as prefix here. IMO all members of a device descriptor are internal and the user should treat its contents as opaque anyway, so no need to specifically mark them as internal. (E.g. if the user indeed would be allowed to modify other members, IMO there should be functions for accessing them added. This allows to change the data structure easier. You can use `static inline` functions where appropriate, so that this would not add any overhead.)

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+/**
+ * @brief   Initialize the given modbus.
+ *
+ * Initialize fields of modbus and UART, RTS pin.
+ *
+ * @param[in] modbus    pointer modbus 
+ *
+ * @return              0 on success
+ * @return              less 0 on error
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:129]: Please use only one `@return` statement. If you want to document different return values, use

```C
 /**
  * @retval 0       On success
  * @retval <0      On error
  */
```

instead. Note that the next "word" after `@retval` is interpreted as value, so there should be no space between `<` and `0` there - otherwise `<` would be interpreted as value `0` as start of the description.

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+ */
+int modbus_rtu_init(modbus_rtu_t *modbus);
+/**
+ * @brief   Send request to slave.
+ *
+ * Send request to slave, blocking and wait response.
+ *
+ * @param[in] modbus    pointer modbus 
+ * @param[in] message    pointer modbus message
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:138]: Nitpick: Please align the parameter and description. (Once you have configured your editor to use spaced over tabs, the tab key will be your friend.)

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ * Initialize fields of modbus and UART, RTS pin.
+ *
+ * @param[in] modbus    pointer modbus 
+ *
+ * @return              0 on success
+ * @return              less 0 on error
+ */
+int modbus_rtu_init(modbus_rtu_t *modbus);
+/**
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:132]: Nitpick: Please add a space after the function definition. (More instances below.)

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ * @param[in] modbus    pointer modbus 
+ * @param[in] message   pointer modbus message
+ *
+ * @return              0 on success
+ * @return              less 0 on error
+ */
+int modbus_rtu_send_response(modbus_rtu_t *modbus, modbus_rtu_message_t *message);
+
+/** @} */
```
Comment (maribu) [inline drivers/modbus_rtu/include/modbus_rtu.h:173]: Nitpick: No `\n` character at the end of this line. (You might want to configure your editor to automatically terminate the last line of every source code file.) The CI tools will enforce that white space errors like this are addressed, as they prevent merging otherwise.

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 8 above)
Comment (theltybc) [reply drivers/modbus_rtu/include/modbus_rtu.h:132]: You mean new line?

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 1 above)
Comment (theltybc) [reply drivers/modbus_rtu/include/modbus_rtu.h]: But this parameter pass to
`int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)`
And I think change and confuse name is bad idea.

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 3 above)
Comment (theltybc) [reply drivers/modbus_rtu/include/modbus_rtu.h:71]: This mean
before transmit set `gpio_write(pin_rts, pin_rts_enable);`
after `gpio_write(pin_rts, !pin_rts_enable);`
Maybe this redundant, but in many library this parameter can be setup.

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 1 above)
Comment (maribu) [reply drivers/modbus_rtu/include/modbus_rtu.h]: But is it really needed to store the symbol rate in the struct? Couldn't it be a parameter to `modbus_rtu_init()` instead? After the initialization, it should no longer be needed (unless I'm missing something). So it would safe 4 bytes of RAM if it is not part of the struct.

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 4 above)
Comment (theltybc) [reply drivers/modbus_rtu/include/modbus_rtu.h:75]:
``` C
typedef struct {
  ...
  modbus_rtu_message_t *msg;
  kernel_pid_t pid;
  uint8_t id;
  uint8_t size_buffer;
} modbus_rtu_t;
```
Means reorder like this?

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 3 above)
Comment (maribu) [reply drivers/modbus_rtu/include/modbus_rtu.h:71]: You could add

```C
static inline int modbus_rtu_pin_enabled(modbus_rtu_t *modbus)
{
    return (modbus->pin_rts == GPIO_UNDEF) ? 0 : 1;
}
```

And than use `modbus_rtu_pin_enabled(modbus)` instead of `modbus->pin_rts_enable`. To me, this is equally readable, but the former safes 4 bytes of RAM.

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 4 above)
Comment (maribu) [reply drivers/modbus_rtu/include/modbus_rtu.h:75]: I have to see the whole struct to confirm this, but the part shown looks good.

The problem I'm referring to is best illustrated with an example. Let's say we have a structure `stuct foo1` declared like this:

```C
struct foo1 {
    uint8_t a;
    uint32_t b
    uint8_t c;
    uint8_t d;
    uint8_t e;
};
```

The whole structure has an alignment requirement of 4 bytes, as the maximum of the alignment requirements of its members is 4 bytes. (Member `uint32_t b` has an alignment requirement of 4 bytes, at least on 32 bit ARM CPUs. All other members have an alignment requirement of 1 bytes.) The C compiler will now fill padding bytes in front of every member whose padding requirement is not fulfilled, until it is fulfilled. Finally, padding bytes are added in the end (if needed), until the padding requirement of the whole struct is fulfilled. (This is needed e.g. to allow an array of the struct.)

The actual memory layout would look like this:

```C
struct foo1 {
    uint8_t a;             /* <-- Alignment requirement of 1 byte */
    uint8_t __padding1[3]; /* <-- Three bytes of padding inserted */
    uint32_t b;            /* <-- Alignment requirement of 4 bytes */
    uint8_t c;             /* <-- Alignment requirement of 1 byte */
    uint8_t d;             /* <-- Alignment requirement of 1 byte */
    uint8_t e;             /* <-- Alignment requirement of 1 byte */
    uint8_t __pading2[1];  /* <-- one byte of padding inserted */
};
```

The following `struct foo2` would need no padding at all (and safe 4 bytes because of this), even though it contains the same members:

```C
struct foo2 {
    uint32_t b;
    uint8_t a;
    uint8_t c;
    uint8_t d;
    uint8_t e;
}
```

Alternatively, `struct foo3` would also need no padding:

```C
struct foo3 {
    uint8_t a;
    uint8_t c;
    uint8_t d;
    uint8_t e;
    uint32_t b;
}
```

To sum up: If you order members by alignment requirements, you will always come up with the least amount of padding. Alternatively, you can carefully order members yourself and also end up with the least amount of padding bytes, but possible with a more "sensible" order regarding what the members express. You can use whatever approach is more appealing to you. It would however be nice, if padding could be reduced to a minimum to safe same RAM. (In the end, we will access members by name anyway, so the order is not so relevant to us, but it is relevant to the size of the `struct`.)

Comment (theltybc) [reply drivers/modbus_rtu/include/modbus_rtu.h:75]:
```C
typedef struct {
  gpio_t pin_rts; // unsigned
  int pin_rts_enable;
  uint32_t timeout;
  uint32_t rx_timeout;
  modbus_rtu_message_t *msg;
  mutex_t mutex_buffer; // list_node_t*
  uart_t uart; // unsigned
  kernel_pid_t pid; // uint16_t
  uint8_t id;
  uint8_t size_buffer;
  uint8_t buffer[MODBUS_RTU_PACKET_SIZE_MAX];
} modbus_rtu_t;
```
Full structure

File: drivers/modbus_rtu/modbus_rtu.c [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+  NB_HI,   //!< Number of coils or registers high byte
+  NB_LO,   //!< Number of coils or registers low byte
+  BYTE_CNT //!< byte counter
+};
+
+// bytes with crc16
+#define MIN_SIZE_REQUEST (8)
+
+#define lowByte(w) ((uint8_t)((w)&0xff))
+#define highByte(w) ((uint8_t)((w) >> 8))
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:33]: Maybe you can use something from the `core/include/byteorder.h` library instead?

File: drivers/modbus_rtu/modbus_rtu.c [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+  while (modbus->size_buffer < num) {
+    if (xtimer_msg_receive_timeout(&msg, modbus->rx_timeout) < 0) {
+      return -1;
+    }
+  }
+  return 0;
+}
+// from buffer
+static inline void read_address(modbus_rtu_t *modbus) {
+  modbus->msg->addr = ((modbus->buffer[ADD_HI]) << 8) | modbus->buffer[ADD_LO];
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:55]: `(modbus->buffer[ADD_HI]) << 8` something tells me that this may just be bit shifting an 8 bit number.
```suggestion
  modbus->msg->addr = byteorder_bebuftohs(&modbus->buffer[ADD_HI]);
```
if you use the `byteorder.h` library.

File: drivers/modbus_rtu/modbus_rtu.c [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+}
+// to buffer
+static inline void write_address(modbus_rtu_t *modbus) {
+  modbus->buffer[ADD_HI] = highByte(modbus->msg->addr);
+  modbus->buffer[ADD_LO] = lowByte(modbus->msg->addr);
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:60]:
```suggestion
  byteorder_htobebufs(&modbus->buffer[ADD_HI], modbus->msg->addr);
```

File: drivers/modbus_rtu/modbus_rtu.c [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+}
+// from buffer
+static inline void read_count(modbus_rtu_t *modbus) {
+  modbus->msg->count = ((modbus->buffer[NB_HI]) << 8) | modbus->buffer[NB_LO];
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:64]: use `byteorder.h`

File: drivers/modbus_rtu/modbus_rtu.c [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+}
+// to buffer
+static inline void write_count(modbus_rtu_t *modbus) {
+  modbus->buffer[NB_HI] = highByte(modbus->msg->count);
+  modbus->buffer[NB_LO] = lowByte(modbus->msg->count);
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:69]: use `byteorder.h`

File: drivers/modbus_rtu/modbus_rtu.c [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+}
+
+static void copy_bits(uint8_t *dst, uint16_t start_bit_dst,
+                      const uint8_t *src, uint16_t start_bit_src, uint16_t count) {
+  while (count > 0) {
+    div_t s = div(start_bit_src, 8);
+    div_t d = div(start_bit_dst, 8);
+    // printf("dst byte: %d; bit: %d\n", d.quot, d.rem);
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:77]: if you want to have debug prints try using the `DEBUG()` or `DEBUG_PRINT()` macro.

See `debug.h` for more info or even `log.h` to be very fancy.

File: examples/modbus/Makefile [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+# This has to be the absolute path to the RIOT base directory:
+RIOTBASE ?= $(CURDIR)/../..
+
+# Comment this out to disable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+DEVELHELP ?= 1
+
+FEATURES_USED += gpio
```
Comment (MrKevinWeiss) [inline examples/modbus/Makefile:15]: Do you need these if they are in the Makefile.deps (periph_gpio and xtimer)

File: drivers/modbus_rtu/modbus_rtu.c [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+    gpio_write(modbus->pin_rts, !modbus->pin_rts_enable);
+  }
+  mutex_init(&(modbus->mutex_buffer));
+  // usec in sec / byte per second * 1.5
+  modbus->rx_timeout = 1000000 / (baudrate / 10) * 1.5;
+
+  modbus->size_buffer = 0;
+  modbus->msg = NULL;
+
+  puts("modbus_init");
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:107]: Maybe better to use the debug or log functions as a user may not want any printout.

File: drivers/modbus_rtu/modbus_rtu.c [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+  int err = uart_init(modbus->uart, baudrate, rx_cb, modbus);
+  if (err != UART_OK) {
+    return err;
+  }
+  if (modbus->pin_rts != GPIO_UNDEF) {
+    gpio_init(modbus->pin_rts, GPIO_OUT);
+    gpio_write(modbus->pin_rts, !modbus->pin_rts_enable);
+  }
+  mutex_init(&(modbus->mutex_buffer));
+  // usec in sec / byte per second * 1.5
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:101]: Please use c comments
```suggestion
  /* usec in sec / byte per second * 1.5 */
```

File: drivers/modbus_rtu/modbus_rtu.c [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+  modbus->rx_timeout = 1000000 / (baudrate / 10) * 1.5;
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:102]: I guess we are not assuming any parity/data/stop bits are change?

File: drivers/modbus_rtu/modbus_rtu.c [hunk 20]
```diff
@@ -0,0 +1,469 @@
...
+  write_address(modbus);
+  uint8_t size;
+  div_t d;
+  switch (modbus->msg->func) {
+  case MB_FC_READ_COILS:
+  case MB_FC_READ_DISCRETE_INPUT:
+  case MB_FC_READ_REGISTERS:
+  case MB_FC_READ_INPUT_REGISTER:
+    write_count(modbus);
+    // id + func + addr + count
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:140]: I guess in the end all the comments will be converted to C style so I will stop nitpicking :)

File: drivers/modbus_rtu/modbus_rtu.c [hunk 21]
```diff
@@ -0,0 +1,469 @@
...
+  modbus->buffer[modbus->size_buffer] = lowByte(crc);
+  modbus->size_buffer++;
+  modbus->buffer[modbus->size_buffer] = highByte(crc);
+  modbus->size_buffer++;
+  enamdle_trasmit(modbus);
+  uart_write(modbus->uart, modbus->buffer, modbus->size_buffer);
+  disamdle_trasmit(modbus);
+}
+
+static inline void enamdle_trasmit(modbus_rtu_t *modbus) {
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:436]: Do you mean enable?

File: drivers/modbus_rtu/modbus_rtu.c [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+  if (modbus->pin_rts != GPIO_UNDEF) {
+    gpio_write(modbus->pin_rts, modbus->pin_rts_enable);
+  }
+}
+
+static inline void disamdle_trasmit(modbus_rtu_t *modbus) {
```
Comment (MrKevinWeiss) [inline drivers/modbus_rtu/modbus_rtu.c:442]: disable?

File: examples/modbus/Makefile [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+USEMODULE += $(filter periph_%,$(FEATURES_USED))
+USEMODULE += modbus_rtu  
+USEMODULE += xtimer
```
[the remark this answers is not in the document]
Comment (theltybc) [reply examples/modbus/Makefile:18]: No need.

File: drivers/modbus_rtu/modbus_rtu.c (hunk 11 above)
Comment (theltybc) [reply drivers/modbus_rtu/modbus_rtu.c:54]: May be better? 
```C
byteorder_bebuftohs(modbus->buffer + ADD_HI);
```

File: drivers/modbus_rtu/modbus_rtu.c (hunk 19 above)
Comment (theltybc) [reply drivers/modbus_rtu/modbus_rtu.c:101]: I`m forgot about this. I'll do it later.

File: drivers/modbus_rtu/modbus_rtu.c (hunk 22 above)
Comment (theltybc) [reply drivers/modbus_rtu/modbus_rtu.c:470]: `enamdle_trasmit`  set `pin_rts` to value `pin_rts_enable`, and `disamdle_trasmit` to value `!enamdle_trasmit`.

Comment (theltybc) [reply drivers/modbus_rtu/modbus_rtu.c:470]: i'm don't know how name this better.

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 2 above)
Comment (theltybc) [reply drivers/modbus_rtu/include/modbus_rtu.h]: New comment  `RTS pin. GPIO_UNDEF if not use.`

File: drivers/modbus_rtu/include/modbus_rtu.h (hunk 3 above)
Comment (theltybc) [reply drivers/modbus_rtu/include/modbus_rtu.h:82]: It used in 2 function. I`m think this not necessary.

File: drivers/modbus_rtu/modbus_rtu.c (hunk 15 above)
Comment (theltybc) [reply drivers/modbus_rtu/modbus_rtu.c]: I will take this into account.

File: drivers/modbus_rtu/modbus_rtu.c (hunk 19 above)
Comment (theltybc) [reply drivers/modbus_rtu/modbus_rtu.c]: Maybe leave these settings with `uart_mode`?

**Comment by theltybc** [comment]
It`s done. 
Check on:
arduino-nano + libmodbus.
blue-pill + libmodbus.
blue-pill +  Iskra Nano Pro (library [Modbus-Master-Slave-for-Arduino](https://github.com/theltybc/Modbus-Master-Slave-for-Arduino))

Code with [libmodbus](https://github.com/stephane/libmodbus/):
    [client](https://github.com/theltybc/c-examples/blob/master/ex/modbus_client.c) 
    [server](https://github.com/theltybc/c-examples/blob/master/ex/modbus_server.c).

**Comment by theltybc** [comment]
I`m done. If you have any offers, you are welcome say it.

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+enum {
+    MB_ER_NONE              = 0,
+    MB_ER_ILLEGAL_FUNCTION  = 1,
+    MB_ER_ILLEGAL_ADDRESS,
+    MB_ER_ILLEGAL_VALUE,
+    MB_ER_SERVER_FAILURE,
+    MB_ER_ACKNOWLEDGE,
+    MB_ER_SERVER_BUSY,
+    MB_ER_MEMORY_PARITY_ERROR,
+    /* non standart */
```
Comment (basilfx) [inline drivers/modbus_rtu/include/modbus_rtu.h:81]:
```suggestion
    /* non standard */
```

File: examples/modbus/Makefile [hunk 25]
```diff
@@ -0,0 +1,47 @@
+# name of your application
+APPLICATION = hello-world
+
+# If no BOARD is found in the environment, use this default:
+BOARD ?= bluepill
+# BOARD ?= arduino-nano
```
Comment (basilfx) [inline examples/modbus/Makefile:6]: Remove if not needed.

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+ * @param[in] modbus    pointer modbus
+ * @param[in] message   pointer modbus message
+ *
+ * @return              0 on success, otherwise error
+ */
+int modbus_rtu_send_response(modbus_rtu_t *modbus,
+                             modbus_rtu_message_t *message);
+
+/**
+ * @brief   Copy number bit from src to dst.
```
Comment (basilfx) [inline drivers/modbus_rtu/include/modbus_rtu.h:214]:
```suggestion
 * @brief   Copy number of bits from src to dst.
```

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 27]
```diff
@@ -0,0 +1,231 @@
...
+ * @return              0 on success, otherwise error
+ */
+int modbus_rtu_send_request(modbus_rtu_t *modbus,
+                            modbus_rtu_message_t *message);
+
+/**
+ * @brief   Wait request from master.
+ *
+ * Wait request from master, blocking.
+ * Request will be writeт to message.
```
Comment (basilfx) [inline drivers/modbus_rtu/include/modbus_rtu.h:188]:
```suggestion
 * Request will be written to message.
```

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+ * Use only slave.
```
Comment (basilfx) [inline drivers/modbus_rtu/include/modbus_rtu.h:189]:
```suggestion
 * Use as slave only.
```

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+ *
+ * @param[in] modbus    pointer modbus
```
Comment (basilfx) [inline drivers/modbus_rtu/include/modbus_rtu.h:191]:
```suggestion
 * @param[in] modbus    pointer to modbus
```

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 30]
```diff
@@ -0,0 +1,231 @@
...
+ * @param[in] message   pointer modbus message
```
Comment (basilfx) [inline drivers/modbus_rtu/include/modbus_rtu.h:192]:
```suggestion
 * @param[in] message   pointer to modbus message
```

File: examples/modbus/Makefile [hunk 31]
```diff
@@ -0,0 +1,47 @@
+# name of your application
+APPLICATION = hello-world
```
Comment (basilfx) [inline examples/modbus/Makefile:2]:
```suggestion
APPLICATION = modbus
```

File: examples/modbus/Makefile [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+# CFLAGS += "-D MODBUS_RTU_USE_WRITE_REGISTER"
+# CFLAGS += "-D MODBUS_RTU_USE_WRITE_COILS"
+# CFLAGS += "-D MODBUS_RTU_USE_WRITE_REGISTERS"
+
+# Change this to 0 show compiler invocation lines by default:
+QUIET ?= 1
+
+include $(RIOTBASE)/Makefile.include
+
+LOADER=$(RIOTBASE)/dist/tools/jlink/jlink.sh
```
Comment (basilfx) [inline examples/modbus/Makefile:33]: Are these lines (until below) necessary?

File: examples/modbus/main.c [hunk 33]
```diff
@@ -0,0 +1,293 @@
...
+
+#include "modbus_rtu.h"
+#include "periph/gpio.h"
+#include "thread.h"
+#include "xtimer.h"
+#include "msg.h"
+#define ENABLE_DEBUG (1)
+#include "debug.h"
+
+
```
Comment (basilfx) [inline examples/modbus/main.c:13]: Remove whitespace.

File: drivers/modbus_rtu/include/modbus_rtu.h [hunk 34]
```diff
@@ -0,0 +1,231 @@
...
+     * @brief   UART port.
+     */
+    uart_t uart;
+    /**
+     * @brief   PID thread. **Must never be changed by the user.**
+     * @internal
+     */
+    kernel_pid_t pid;
+    /**
+     * @brief   ID device. 0 - master, otherwice slave.
```
Comment (basilfx) [inline drivers/modbus_rtu/include/modbus_rtu.h:142]:
```suggestion
     * @brief   ID of the device, 0 is reserved for master.
```

**Review by basilfx — CHANGES_REQUESTED** [review]

File: examples/modbus/Makefile (hunk 32 above)
Comment (theltybc) [reply examples/modbus/Makefile:33]: While checking. After that it will be deleted.
